### DISCUSSION TITLE
Is it possible to build Go binaries for Windows?

### DISCUSSION THREAD

**Discussion by Darthagnon** [photoview/photoview#127] [General]
I've seen a fair amount of GoLang projects with binaries for lots of operating systems, and e.g. for Windows servers, it's much better than Docker. 

Is it possible to compile this project to a binary (in my case, for Windows)? Not sure which file to point the compiler to.

**Comment by viktorstrate** [comment]
Yes that is absolutely possible, just navigate to the `/api` directory and execute

    $ go build -v -o photoview .

But to run the server without docker, you would need to configure a database and how you want to serve the web iterface manually aswell.

I suggest you copy `api/example.env` to `.env` and change the environment variables from there accordingly.

If you prefer, you can serve the UI yourself, eg. with apache or nginx, by building the ui with the correct API_ENDPOINT environment variable. Otherwise you can use the go binary to serve the web files by enabling SERVE_UI and copying the builded ui files to a directory called `./ui` located next to the binary.

If you need more information, I strongly suggest you to look at the [Dockerfile](https://github.com/viktorstrate/photoview/blob/master/Dockerfile) to see how it configures everything.
You can also look at [setting up a development environment](https://github.com/viktorstrate/photoview#setup-development-environment), as it's very similar to building it for production.

Please not though, that I haven't tested Photoview on Windows, so I don't know if that works.

**Comment by Darthagnon** [comment]
Thank you for the detailed guide! I'll try building it this weekend, and post back with progress (and binaries packages if successful).

Double-checking, ```SERVE_UI``` is an env var that should be set to ```1```?

**Comment by viktorstrate** [comment]
Awesome, let me know how it goes.
Yes, SERVE_UI=1 will enable the static file server

**Comment by viktorstrate** [comment]
Also see this issue https://github.com/viktorstrate/photoview/issues/70#issuecomment-676255190

**Comment by Darthagnon** [comment]
I successfully managed to build the server, following your instructions. (not 100% sure about the UI - used what came with the repo; I have NPM installed, so could try running a build of the folder if necessary, though it seemed to already have all the HTML, JS, CSS, etc.). It launches on Windows - looks very promising!

I have not, however, managed to successfully test it, as I couldn't get it to successfully connect to installed MySQL. (Wondering if I should continue this issue over on #130, as that is more about databases, and this one is more about building). I have attached the Windows binary package in a 7z (renamed to zip) file, and a text file with the errors I saw.
[Photoview_report.txt](https://github.com/viktorstrate/photoview/files/5626188/Photoview_report.txt)
[photoview server.zip](https://github.com/viktorstrate/photoview/files/5626189/photoview.server.zip)

**Comment by viktorstrate** [comment]
It looks like it does connect to the database, but it fails to do the initial database setup.
What sql server and version are you using?

For the UI, you should build it before it can be used, the html, js, css files are not bundled and cannot be used by the browser directly. You should be able to build it with `npm run build`, then everything you need will be in a new folder called `dist`.

**Comment by Darthagnon** [comment]
I was using [MySQL Community Server 8.0.22](https://dev.mysql.com/downloads/mysql/). Is there a recommended database server/version to use?

I have now run a build of the GUI (updated server zip attached), and also typed up all your instructions for building on Windows.
[photoview-server-win-1.1.3.zip](https://github.com/viktorstrate/photoview/files/5647931/photoview-server-win-1.1.3.zip)
[Build - Copy.txt](https://github.com/viktorstrate/photoview/files/5647933/Build.-.Copy.txt)

**Reply by DevilsDesigns** [reply]
So everytime i attemp to install with the current install. Ive been trying to build on WLS2 and on Windows 10 Pro. i keep getting this error.

```
Note: This command was run via npm module 'win-node-env'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! photoview-ui@0.0.1 build: `NODE_ENV=production node build.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the photoview-ui@0.0.1 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\DemonWarrior\AppData\Roaming\npm-cache\_logs\2021-03-24T23_29_36_599Z-debug.log
PS C:\Users\DemonWarrior\photoview\ui>

```
ive been attempting to correct this error for over 3 days. Did you get this error and how could i fix this.
Building terminal
https://pastebin.com/vFAGMbxW
Full Build Log
https://pastebin.com/9NEqMy9t

**Comment by viktorstrate** [comment]
Okay, MySQL should be fine, though I haven't tested it myself.
Would you mind trying it with the latest version of MariaDB, that is what the default docker compose config uses.

https://mariadb.org/download/

**Comment by Darthagnon** [comment]
I managed to connect to MariaDB, but it didn't manage to serve the content to the webpages:

http://localhost:4001/api:
```CreateFile \ui\C:\api: The filename, directory name, or volume label syntax is incorrect.```

I tried changing the folder structure, so all the files and folders were in an ```/api``` folder (the way it's structured by default), but it didn't make any difference. Test was run with the last build I posted, above.

http://localhost:1234:
no response; didn't load

CMD Log:
```
2020/12/08 14:50:26 Connecting to database: root:password@tcp(localhost:3306)/test?multiStatements=true&parseTime=true
2020/12/08 14:50:26 Database is up to date
2020/12/08 14:50:26 Initializing scanner queue with 3 workers
2020/12/08 14:50:26 Queue waiting
2020/12/08 14:50:26 Periodic scan interval changed: disabled
2020/12/08 14:50:26 Scan interval runner: Waiting for signal
2020/12/08 14:50:26 Scan interval runner: New ticker detected
2020/12/08 14:50:26 Scan interval runner: Waiting for signal
2020/12/08 14:50:26 Executable worker not found: darktable
2020/12/08 14:50:26 Executable worker not found: ffmpeg
2020/12/08 14:50:26 ?? Graphql playground ready at http://localhost:4001/api
2020/12/08 14:50:54 Did not find auth-token cookie
2020/12/08 14:50:54 [0;34;49mGET [0;31;49m500[0m localhost:4001/api [0;36;49m257.1814ms[0m unauthenticated
2020/12/08 14:50:55 Did not find auth-token cookie
2020/12/08 14:50:55 [0;34;49mGET [0;31;49m500[0m localhost:4001/favicon.ico [0;36;49m1.0012ms[0m unauthenticated
```

Thank you so much for the help you've given me, though, and for listening to my rambling and feature requests! Must say, I'm really looking forward to SQLite if it works out!

**Comment by viktorstrate** [comment]
No problem, I'm a little curious if it will run on Windows as I haven't tested that.
The database migration is comming along nicely.

It looks like the server is running in development mode, make sure the environment variable DEVELOPMENT is not set to 1

**Comment by Darthagnon** [comment]
I noticed the readme now mentions serving the GUI with NPM. I wasn't doing that before (I had assumed that the GoLang exe served everything, the JS, CSS, HTML, etc. like [filebrowser](https://github.com/filebrowser/filebrowser) ); maybe that's why it wasn't working... gotta try again.
